Initialize listeners via NewListeners function

hsjoberg opened the pull request:
This PR is related to issue [lnd#4492](https://github.com/lightningnetwork/lnd/issues/4492) and aims to fix lnd restarting issues.
Currently, the in-mem listeners are in the global scope and initialized once. When the gRPC servers goes down because of a `stopDaemon` call, these listeners will cease to function, meaning we cannot start lnd again, thus we need to (re-)initialize them in the `Start` function. 

This is a breaking change and new code in the lnd repository (coming soon) is required for this to work,  hence the version bump.

halseth commented on templates.go:
```diff
@@ -66,42 +66,44 @@ import (
-var (
-	// {{$lis}} is a global in-memory buffer that listeners that is
+type {{$lis | UpperCase }} struct {
+	// listener is a global in-memory buffer listener that is
```
update comment as it is no longer global

hsjoberg replied:
Good catch, fixed

halseth commented on templates.go:
```diff
@@ -66,42 +66,44 @@ import (
+	optionsMtx sync.Mutex
+}
+
+var {{$lis}} {{$lis | UpperCase}}
```
This will still be global. 

I think we rather should make the `NewListeners` method return a new struct that holds the listeners we need, and then we can pass these listeners to the methods that use them.

hsjoberg replied:
@halseth That definitely makes sense.

Is it possible though? Inside [`bindings.go`](https://github.com/lightningnetwork/lnd/blob/b4bf4b2906c066ec0d0b8d7183c720b1d7d19220/mobile/bindings.go#L71-L80) they could be passed to the lnd conf so that the listeners get properly set up in lnd's gPRC side of things.

However we also have the actual generated gRPC functions that are directly called from the host application.
These functions are currently calling `get<listener>Conn()`:
https://github.com/lightninglabs/falafel/blob/1bfcffa7997475acca5c0ad10100fb6001dd7d55/templates.go#L39-L40

I don't see how we could prevent some kind of global scope as we cannot pass the listeners back to the host, thus we must set a state inside falafel's gRPC files.

halseth replied:
Ah, I see. The generated gRPC functions are parsed by gomobile and converted to Java or Objective-C compatible bindings, and they are currently "statitc", in that they reference the running LND through the global variables.

To achieve what I initially thought, we would need these methods to reference a local LND instance, so we could all them like this:
```go
func (l *Lnd) GetInfo(...) (...) {

}
```
 I haven't tried this so not sure if it even is supported by gomobile passing the LND reference across the FFI boundary like that. If not, then I think you are right that it would be easier just sticking to globals.

hsjoberg replied:
Yes precisely.

I'm not sure that passing the LND reference across the FFI boundary is even preferable, I don't have any particular argument against it but it sounds like an unsound idea.

Perhaps this could be addressed a later point as well.
My suggestion is to go with the solution we have currently. It's still an improvement albeit not quite perfect.

halseth replied:
Yep, agreed. It is a step in the right direction 👍

halseth commented on main.go:
```diff
@@ -15,7 +15,7 @@ import (
 
 const toolName = "falafel"
-const version = "0.7.1"
+const version = "0.8.0"
```
👍

halseth reviewed (changes requested):
Thanks for the PR, first contributor PR to falafel! 🎉 

I think you are on the right track, left some suggestions.

hsjoberg commented:
I opened a PR in the lnd repository fixing the lnd mobile package to work with the code purposed here. :slightly_smiling_face: 
https://github.com/lightningnetwork/lnd/pull/4599

halseth commented:
Probably best to check how it works on top of #9

hsjoberg commented:
@halseth Okay, I've changed it back to version 0.8.0.
I haven't had time to check with #9 yet.

halseth reviewed (changes requested):
Looked at it again now, and it will conflict with #9 since we move the dial options to the service instead of the listeners now.

After rebasing this PR will probably become even simpler§ 😁

hsjoberg commented:
I took a look at #9, but I don't understand it enough to be able to base this PR upon it. So I would prefer if we could merge this one first.

hsjoberg commented:
> After rebasing this PR will probably become even simpler§ 😁

@halseth Ah I see, yeah it was easier to understand the changes by looking at the actual generated files, and yes not much has to be changed now.

I've rebased and updated the PR.
It has _not_ been tested though because of the mobile-situation in current lnd master.
I'll try it out together with https://github.com/lightningnetwork/lnd/pull/5010 etc when I got time.

halseth commented on templates.go:
```diff
@@ -69,6 +69,12 @@ var (
+func NewListeners() {
+{{range $lis := .Listeners}}
+	{{$lis}} = bufconn.Listen(100)
+{{- end}}
```
cool, I didn't know of this syntax :) Looks like we can also use `- range` above to avoid the leading newline 😁

hsjoberg replied:
Oh yeah, I'll fix it on the other place as well to keep it consistent. 🙂

hsjoberg replied:
Fixed.

halseth commented on templates.go:
```diff
@@ -50,7 +50,7 @@ var (
 	// referenced by the generated mobile APIs, such that all client calls
 	// will be going through it.
-	{{$lis}} = bufconn.Listen(100)
+	{{$lis}} *bufconn.Listener
```
To keep this backwards compatible, we could just let this stay the same, then make the method below called `RecreateListeners`. That way it's not mandatory to call it.

hsjoberg replied:
Perhaps, but IIUC lnd as of right now in master is already incompatible falafel until https://github.com/lightningnetwork/lnd/pull/5010 is merged.

I would suggest we keep it as is and work on getting https://github.com/lightningnetwork/lnd/pull/4599 merged.

hsjoberg replied:
One possible solution could be to listen for the shutdown signal in a goroutine and then `RecreateListeners` there.

halseth replied:
> Perhaps, but IIUC lnd as of right now in master is already incompatible with falafel until [lightningnetwork/lnd#5010](https://github.com/lightningnetwork/lnd/pull/5010) is merged.

Sure, but I see no downside on having the listeners be created by default? (without first calleing `NewListeners`)

hsjoberg replied:
@halseth Yeah it's mostly a question on how and where it would be reinitialized then. I can change it to `RecreateListeners` and then use the shutdown channel for recreation.

Then something like this in lnd's mobile code:

```diff
@@ -134,4 +134,11 @@ func Start(extraArgs string, unlockerReady, rpcReady Callback) {
 
                rpcReady.OnResponse([]byte{})
        }()
+
+       go func() {
+               <-shutdownInterceptor.ShutdownChannel()
+               // When lnd and the gRPC server goes down, it will close our listeners.
+               // So we have to re-initialize them for the next run
+               RecreateListeners()
+       }()
 }
```

halseth replied:
Ah, my idea was that you call it the same way as you are using `NewListeners` now.

hsjoberg replied:
@halseth Sure that works too, I guess it costs next to nothing doing that, just a bit ugly. 😛

halseth commented on templates.go:
```diff
@@ -69,6 +69,12 @@ var (
 	serviceDialOptionsMtx sync.Mutex
 )
 
+func NewListeners() {
```
method should have a godoc.

hsjoberg replied:
Oops! Fixed, thanks.

halseth reviewed (changes requested):
Thanks, I've tested this and it works as advertised 🤠  Left a few style suggestions, when that is addressed we can merge this.

halseth commented:
I went ahead and did the last small fixups I saw and committed this as https://github.com/lightninglabs/falafel/commit/58e52ea497a7e130ef895399a6e8239a31ec9932

I also tagged and pushed a new version `v0.9.0`

Thanks!

hsjoberg commented:
@halseth Awesome, thanks!
